eLabFTW integration via Galaxy file source

domgz opened the pull request:
Closes #18665, requires #19154 and #19256 (and also includes them, sorry; to review have a look at [c00d250](https://github.com/galaxyproject/galaxy/pull/19319/commits/c00d250a225e5ce83475c1e92d30373160257c10)). Implements a file source to integrate Galaxy with [eLabFTW](https://www.elabftw.net/).

[eLabFTW](https://www.elabftw.net/) revolves around the concepts of [experiment](https://doc.elabftw.net/user-guide.html#experiments) and [resource](https://doc.elabftw.net/user-guide.html#resources). Experiments and resources can have files attached to them. To get a quick overview, try out the live demo at [demo.elabftw.net](https://demo.elabftw.net). The scope of this implementation is exporting data from and importing data to eLabFTW as file attachments of already existing experiments and resources. Each user can configure their preferred eLabFTW instance entering its URL and an API Key.

File sources reference files via a URI, while eLabFTW uses auto-incrementing positive integers. For more details read galaxyproject/galaxy#18665. This leads to the need to declare a mapping between said identifiers and Galaxy URIs.

Those take the form `elabftw://demo.elabftw.net/entity_type/entity_id/attachment_id`, where:
- `entity_type` is either 'experiments' or 'resources'
- `entity_id` is the id (an integer in string form) of an experiment or resource
- `attachment_id` is the id (an integer in string form) of an attachment

This implementation uses both `aiohttp` and the `requests` libraries as underlying mechanisms to communicate with eLabFTW via its [REST API](https://doc.elabftw.net/api/v2). A significant limitation of the implementation is that, due to the fact that the API does
not have an endpoint that can list attachments for several experiments and/or resources with a single request, when
listing the root directory or an entity type _recursively_, a list of entities has to be fetched first, then to fetch
the information on their attachments, a separate request has to be sent _for each one_ of them. The `aiohttp` library makes it bearable to recursively browse instances with up to ~500 experiments or resources with attachments by sending them
concurrently, but ultimately solving the problem would require changes to the API from the eLabFTW side.

This is the third and last PR of a series of PRs that integrate eLabFTW with Galaxy via a file source (together they address issue galaxyproject/galaxy#18665):
- galaxyproject/galaxy#19154
- galaxyproject/galaxy#19256
- galaxyproject/galaxy#19319

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [x] Instructions for manual testing are as follows:
  1. Run eLabFTW, for example [using Docker Compose](https://github.com/elabftw/elabimg/blob/master/src/docker-compose.yml-EXAMPLE).
  2. Copy the configuration samples from file_sources_conf.yml.sample and user_preferences_extra_conf.yml.sample to your own configuration files.
  3. Create an account and generate an API Key on eLabFTW.
  4. Configure the endpoint and API Key on the user preferences page.
  5. Create an experiment or resource in eLabFTW and attach a file.
  6. Import the file to your history from the eLabFTW file source.
  7. Export your history to the eLabFTW file source.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

domgz committed:
Support listing file sources asynchronously

The method `list()` from `galaxy.files.sources.BaseFilesSource` lists the directories and files within a file source. An optional keyword argument `recursive` (`False` by default) lets it recursively retrieve directories and files within a specific directory.

This operation is very cheap in terms of CPU and expensive in IO terms, be it network or filesystem IO. Depending on how the underlying system is built, it may support retrieving directories and files recursively or not. If it does not, then every time a directory is listed, it is necessary to make another request to list each subdirectory. This may end up involving hundreds of requests. Done sequentially, this can be extremely slow, especially if each one involves network access.

This commit makes the `list()` method asynchronous, which enables Galaxy to wait for the underlying system to complete the requests concurrently, resulting in a massive speedup. The price to pay is the extra complexity of using the async primitives.

Since this change implies that all functions in the chain up to the API endpoints and the test functions must also be made asynchronous, this commit also takes care of it.

domgz committed:
Implement a file source to integrate Galaxy with eLabFTW

eLabFTW [1] revolves around the concepts of experiment [2] and resource [3]. Experiments and resources can have files attached to them. To get a quick overview, try out the live demo [4]. The scope of this implementation is exporting data from and importing data to eLabFTW as file attachments of already existing experiments and resources. Each user can configure their preferred eLabFTW instance entering its URL and an API Key.

File sources reference files via a URI, while eLabFTW uses auto-incrementing positive integers. For more details read galaxyproject/galaxy#18665 [5]. This leads to the need to declare a mapping between said identifiers and Galaxy URIs.

Those take the form `elabftw://demo.elabftw.net/entity_type/entity_id/attachment_id`, where:
- `entity_type` is either 'experiments' or 'resources'
- `entity_id` is the id (an integer in string form) of an experiment or resource
- `attachment_id` is the id (an integer in string form) of an attachment

This implementation uses both `aiohttp` and the `requests` libraries as underlying mechanisms to communicate with eLabFTW via its REST API [6]. A significant limitation of the implementation is that, due to the fact that the API does not have an endpoint that can list attachments for several experiments and/or resources with a single request, when listing the root directory or an entity type _recursively_, a list of entities has to be fetched first, then to fetch the information on their attachments, a separate request has to be sent _for each one_ of them. The `aiohttp` library makes it bearable to recursively browse instances with up to ~500 experiments or resources with attachments by sending them concurrently, but ultimately solving the problem would require changes to the API from the eLabFTW side.

References:
- [1] https://www.elabftw.net/
- [2] https://doc.elabftw.net/user-guide.html#experiments
- [3] https://doc.elabftw.net/user-guide.html#resources
- [4] https://demo.elabftw.net
- [5] https://github.com/galaxyproject/galaxy/issues/18665
- [6] https://doc.elabftw.net/api/v2

domgz committed:
Mark tests using `assert_list_names` as async

domgz committed:
Merge branch 'file_sources_assign_paths' into eLabFTW

domgz committed:
Merge branch 'file_sources_asynchronous_listing' into eLabFTW

domgz committed:
Merge branch 'dev' into eLabFTW

davelopez reviewed (commented):
As I mentioned in person, the implementation looks amazing! especially because of the care and attention to detail! Thank you very much!

I also think that the `recursive` option that has motivated many of these changes (and increased complexity) is a use case that we should probably not support. Being able to recursively list or select a potentially huge number of experiments without strict pagination will likely be unusable or cause more trouble than it is worth... :sweat:

mvdbeek commented on lib/galaxy/files/sources/__init__.py:
```diff
@@ -296,7 +299,7 @@ def get_uri_root(self) -> str:
 
     @abc.abstractmethod
-    def list(
+    async def list(
```
Please don't change this for existing interfaces. You can add a `<operation>_async` method where needed. We can't do blocking IO in async calls, but every plugin apart from yours will make blocking calls.

domgz replied:
Thanks a lot for mentioning this, I was not aware of the inner workings of FastAPI and it would have been a big mistake to merge this PR as it is. I reverted commits from #19256 (so the endpoint is synchronous again), then chose a different mechanism to send the requests concurrently (see 204b5fe).

Namely, I learned that while FastAPI runs async routes on the main thread's event loop (that's precisely the reason why the previous approach is a bad idea), it runs sync routes on a separate thread from a threadpool. Therefore, the alternative I chose is to spawn an event loop on the sync route's thread, send the concurrent requests using it, and close it afterward.

This implies that #19256 would no longer be needed.

mvdbeek replied:
Thanks. I guess my other question is, have you checked with the elab people that concurrent requests are not considered abuse ? They often are in other contexts, e.g. in galaxy we would ban your ip or you would be hit by rate limits.

domgz replied:
**TL;DR**: _no public instances thus hard to define abuse and to be abused, cannot be engineered much better without eLabFTW API changes_

eLabFTW's scope is one instance per lab or at most one per research institution. Unlike Galaxy, there are no public main instances. Either you host your own, or you use their [hosted offer](https://www.deltablot.com/elabftw/), which I assume is a single instance just for you (it costs 4985 € / yr) and is supposed to handle _"~128 active users max"_.

Therefore, there is no authority defining abuse that it makes sense to ask or that would ban your ip per se. It depends on how many resources the endpoint you have configured has at their disposal, what their admins consider reasonable, and whether they enforce rate limits or not.

Nevertheless, and foreseeing this problem, I included a hard limit [`MAX_CONCURRENT_REQUESTS: int = 75`](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R88), setting the number to something that would not cause issues (for example timed out requests according to the [timeouts I also have defined](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R90-R92)) on their [live demo site](https://demo.elabftw.net). To decrease the number of requests, I also avoid making requests if it is known that the experiment or resource has no attachments [`wrapped_entity.source["has_attachment"]`](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R459).

From a software engineering standpoint, given the fact that the file sources API requires the plugin to be able to list everything on the server (`recursive=True`) and the pagination is client-side I guess I cannot do much better besides a global queue with a limit that applies to the sum of requests per endpoint from all users at any time.

Achieving a better result if the plugin is supposed to work in the case `recursive=True` [would involve changes to the eLabFTW API](https://github.com/galaxyproject/galaxy/pull/19319/files#diff-226b24a98914f99eb1b630e5ddb06da305bb99bc19386adcd5e538cb1c595be3R18-R24), although that is out of our reach. At best an endpoint [like this one](https://doc.elabftw.net/api/v2/#/User%20Uploads/read-user-uploads) but that lists all attachments _a user can access_, ideally takes offset, limit, sorting and the filtering parameters we need.

davelopez replied:
That is why I think just not supporting `recursive=True` in those file sources that don't make sense would be the safest option. We could add a "support" flag, similar to what we already have for pagination, etc. here:

https://github.com/galaxyproject/galaxy/blob/b1217afc0cd42a280dfadcf6898d50b28c352e25/lib/galaxy/files/sources/__init__.py#L84

Then we can disable the recursive option in the UI and the API according to this flag.

domgz committed:
Revert "Mark tests using `assert_list_names` as async"

This reverts commit a05d5cc5ec53dcf8d696c51b47ca320eb8071c6f.

domgz committed:
Revert "Support listing file sources asynchronously"

This reverts commit 668437aff7bce0a475e55879bbe29e314dfcd8f1.

domgz committed:
List eLabFTW endpoint contents on a separate event loop within the sync route thread

FastAPI runs sync routes on a separate thread (from a threadpool), and async routes on the main thread's event loop. Originally, it was conceived to convert `_list` to an async method (see https://github.com/galaxyproject/galaxy/pull/19256). However, given that all other file source plugins are blocking, they could block the main thread for a significant amount of time (see https://github.com/fastapi/fastapi/issues/3091). Thus, the implementation below creates a new event loop within the sync route's thread so that the eLabFTW plugin can still send concurrent requests without blocking the main thread.

domgz committed:
Merge branch 'dev' into eLabFTW

davelopez commented on lib/galaxy/files/sources/elabftw.py:
```diff
@@ -0,0 +1,990 @@
+        if limit is not None:
+            wrapped_entries = wrapped_entries[:limit]
+
+        return (entries := [wrapped_entry.entry for wrapped_entry in wrapped_entries]), len(entries)
```
This seems to be returning just the paginated count, so it will always display 1 page. For the pagination to work correctly, the number returned in the second position of the tuple should be the `total count` of elements that match the current filter.

domgz replied:
Bad news, I do not see anything in the [API](https://doc.elabftw.net/api/v2/) that provides the total count of elements :sweat_smile:

I checked the response headers too, but no luck (the API key in the example below has been already deleted)

```
curl -v -X 'GET'   'https://demo.elabftw.net/api/v2/experiments'   -H 'accept: application/json'   -H 'Authorization: 3-38795688babb7a6c359ec14cb87b66da8ff32576aced34f5aa28fc7ef695065c82539e45fe06a910bbf73' | jq
Note: Unnecessary use of -X or --request, GET is already inferred.
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0* Host demo.elabftw.net:443 was resolved.
* IPv6: (none)
* IPv4: 163.172.146.126
*   Trying 163.172.146.126:443...
* ALPN: curl offers h2,http/1.1
} [5 bytes data]
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
} [512 bytes data]
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: none
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Server hello (2):
{ [122 bytes data]
* TLSv1.3 (IN), TLS change cipher, Change cipher spec (1):
{ [1 bytes data]
* TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
{ [19 bytes data]
* TLSv1.3 (IN), TLS handshake, Certificate (11):
{ [2032 bytes data]
* TLSv1.3 (IN), TLS handshake, CERT verify (15):
{ [79 bytes data]
* TLSv1.3 (IN), TLS handshake, Finished (20):
{ [52 bytes data]
* TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
} [1 bytes data]
* TLSv1.3 (OUT), TLS handshake, Finished (20):
} [52 bytes data]
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384 / x25519 / id-ecPublicKey
* ALPN: server accepted h2
* Server certificate:
*  subject: CN=demo.elabftw.net
*  start date: Dec 11 18:55:16 2024 GMT
*  expire date: Mar 11 18:55:15 2025 GMT
*  subjectAltName: host "demo.elabftw.net" matched cert's "demo.elabftw.net"
*  issuer: C=US; O=Let's Encrypt; CN=E6
*  SSL certificate verify ok.
*   Certificate level 0: Public key type EC/prime256v1 (256/128 Bits/secBits), signed using ecdsa-with-SHA384
*   Certificate level 1: Public key type EC/secp384r1 (384/192 Bits/secBits), signed using sha256WithRSAEncryption
*   Certificate level 2: Public key type RSA (4096/152 Bits/secBits), signed using sha256WithRSAEncryption
* Connected to demo.elabftw.net (163.172.146.126) port 443
* using HTTP/2
* [HTTP/2] [1] OPENED stream for https://demo.elabftw.net/api/v2/experiments
* [HTTP/2] [1] [:method: GET]
* [HTTP/2] [1] [:scheme: https]
* [HTTP/2] [1] [:authority: demo.elabftw.net]
* [HTTP/2] [1] [:path: /api/v2/experiments]
* [HTTP/2] [1] [user-agent: curl/8.11.1]
* [HTTP/2] [1] [accept: application/json]
* [HTTP/2] [1] [authorization: 3-38795688babb7a6c359ec14cb87b66da8ff32576aced34f5aa28fc7ef695065c82539e45fe06a910bbf73]
} [5 bytes data]
> GET /api/v2/experiments HTTP/2
> Host: demo.elabftw.net
> User-Agent: curl/8.11.1
> accept: application/json
> Authorization: 3-38795688babb7a6c359ec14cb87b66da8ff32576aced34f5aa28fc7ef695065c82539e45fe06a910bbf73
> 
* Request completely sent off
{ [5 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [57 bytes data]
* TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
{ [57 bytes data]
< HTTP/2 200 
< content-type: application/json
< cache-control: max-age=0, private, must-revalidate
< set-cookie: PHPSESSID=0etd4d3088i6032kro36jbsqgh84ft7f7gmhcqp1d6; path=/; secure; HttpOnly; SameSite=Lax
< cache-control: no-cache, private
< date: Fri, 24 Jan 2025 14:55:40 GMT
< strict-transport-security: max-age=63072000
< x-xss-protection: 0
< x-content-type-options: nosniff
< content-security-policy: default-src 'self' data:; script-src 'self' ; connect-src 'self' blob: https://get.elabftw.net; img-src 'self' data: blob:; style-src 'self' 'unsafe-inline'; font-src 'self' data:; object-src 'self'; base-uri 'none'; frame-ancestors 'none'
< referrer-policy: no-referrer
< permissions-policy: autoplay 'none'; camera 'self'; document-domain 'none'; encrypted-media 'none'; fullscreen 'self'; geolocation 'none'; microphone 'self'; midi 'none'; payment 'none'; vr 'none'
< vary: Accept-Encoding
< server: 213
< access-control-allow-credentials: true
< access-control-expose-headers: Location, Content-Encoding, Content-Disposition, Cache-Control
< 
{ [13383 bytes data]
100 30875    0 30875    0     0   124k      0 --:--:-- --:--:-- --:--:--  125k
* Connection #0 to host demo.elabftw.net left intact
[
  {
    "id": 273,
    "team": 1,
    "title": "Electronic Microscopy",
    "date": "2025-01-24",
...
```

domgz replied:
I guess the only thing I can do as workaround is request all items at once then compute the length of the list...

davelopez replied:
Ugh... that is unfortunate indeed... :sweat: 
Then you can try setting `supports_pagination = False` at the plugin level and hope that in most cases this will be connected to a local or institutional instance with a reasonable amount of experiments or resources :sweat_smile:

domgz replied:
I set `supports_pagination = False` in 1d6432e, so now we have "fake" pagination.

![Fake pagination (page 1)](https://github.com/user-attachments/assets/d8741fa9-3262-4d51-82d1-f44109fcdd05)

![Fake pagination (page 2)](https://github.com/user-attachments/assets/d18e6d8e-d099-4db1-96de-8d0c61d92ea3)

bgruening replied:
Oha :)
Please create an upstream issue. Our French colleagues are probably interested in following up on this.

davelopez replied:
Thank you @kysrpex, it should be fine like this in most cases until they add support for a total count query.

domgz replied:
@bgruening https://github.com/elabftw/elabftw/issues/5441

davelopez reviewed (commented):
I've tested and it works nicely!
The only thing I found was a small issue with pagination. Details below.

domgz committed:
Disable pagination for eLabFTW files source

Set `eLabFTWFilesSource.supports_pagination` to `False` to disable pagination. Pagination is not supported because eLabFTW does not return the total count of elements matching a query, see https://github.com/galaxyproject/galaxy/pull/19319#discussion_r1928753352.
